LibWeb/CSS: Parse full `mask` property and its longhands

InvalidUsernameException opened the pull request:
Only parsing, no actual masking functionality is implemented in here.

This gives >400 new WPT subtest passes. Remaining test failures in the imported tests are due to:
* Missing support for `mask-border-*` properties.
* Incorrect behavior of some of the existing data types used by the `mask` properties.

Calme1709 commented on Libraries/LibWeb/CSS/Properties.json:
```diff
@@ -2390,6 +2418,65 @@
+    "inherited": false,
+    "initial": "0% 0%",
+    "valid-types": [
+      "background-position"
```
It would be nice if we could rename `background-position` to `position` as it is in the spec - akin to what you did for `background-repeat` -> `repeat`. Not necessary for this PR so a FIXME would be enough IMO

InvalidUsernameException replied:
Yeah, I was seeing what looked like some other naming inconsistencies. However I only glanced at the spec and then decided not to look into it further since things seems to be inconsistent there as well.

Taking a closer look at this case now, it turns out this was just wrong. `background-position` and `position` are two separate types, the latter's valid values being a subset of the former's. And there is even support for this in the codebase, so I just changed the `valid-types` to `position` here.

This difference doesn't have any behavioral effect though. The parsing code is overwritten in the large `switch` statement in `Parser::parse_css_value()` to support the comma separated list. And the call there was using the right variant of the `position`-parsing function already.

Calme1709 replied:
Looks like I also missed that `background-position` was a distinct type as well.

Just as an FYI - there is a behavioral change here in that we now disallow those `background-position` valid values within the `mask` shorthand, this is because the calls to `parse_css_value_for_properties` works on `valid-types`, bypassing the `parse_a_css_value` switch statement.

Looks good to me now.

AtkinsSJ commented on Libraries/LibWeb/CSS/Properties.json:
```diff
@@ -2432,6 +2432,23 @@
+    "animation-type": "repeatable-list",
+    "inherited": false,
+    "initial": "auto",
+    "max-values": 2,
+    "valid-types": [
+      "length [0,∞]",
+      "percentage [0,∞]"
+    ],
+    "valid-identifiers": [
+      "auto",
+      "cover",
+      "contain"
+    ],
+    "percentages-resolve-to": "length"
```
Not a change to make in this PR, but we could really do with `<bg-size>` being a properly supported type instead of having to specify all its requirements inside `Properties.json`.

AtkinsSJ commented on Libraries/LibWeb/CSS/Properties.json:
```diff
@@ -2370,10 +2370,17 @@
   },
   "mask": {
     "inherited": false,
-    "__comment": "FIXME: add longhands mask-clip, mask-composite, mask-mode, mask-origin, mask-position, mask-repeat, mask-size; also reset mask-border",
```
Can you keep the note about `mask-border`?

AtkinsSJ commented on Libraries/LibWeb/CSS/StyleValues/ShorthandStyleValue.cpp:
```diff
@@ -489,6 +489,109 @@ String ShorthandStyleValue::to_string(SerializationMode mode) const
+            PropertyID canonical_property_order[] = {
+                PropertyID::MaskImage,
+                PropertyID::MaskPosition,
+                // Intentioanlly skipping MaskSize here, it is handled together with MaskPosition.
```
Nit: "Intentionally"

AtkinsSJ commented on Libraries/LibWeb/CSS/StyleValues/ShorthandStyleValue.cpp:
```diff
@@ -489,6 +489,109 @@ String ShorthandStyleValue::to_string(SerializationMode mode) const
+                auto property = canonical_property_order[i];
+                auto const& value = property_value(property);
+
+                if (is_initial_value(property)) {
+                    if (property != PropertyID::MaskOrigin)
+                        continue;
```
I think this will skip serializing `mask-position` if it's the initial value - but that's only safe to do if `mask-size` is also the initial value.

AtkinsSJ reviewed (changes requested):
Thanks for tackling this! Looks good except for one small bug and a couple of nits.

InvalidUsernameException commented:
Relevant changes from addressing the review comments are in the second force-push, which unfortunately isn't visible anymore in the PR history, but corresponds to [this diff](https://github.com/LadybirdBrowser/ladybird/compare/cfc41c47078243004e1b058fd195bbbb1bb88f37..efc066fb795235fa9336953a5f58ca70b5299090).

The others are a rebase and several flaky CI runs.

InvalidUsernameException commented:
I've dropped the aforementioned change to background-size serialization in the last rebase again because the same change was already done in #5677.

Rebasing this on `master` also regressed 5 of subtests from the tests imported for this PR:

https://github.com/LadybirdBrowser/ladybird/blob/ff529d11964dd0681a6ca6e0f4bc74f1d00f3fdb/Tests/LibWeb/Text/expected/wpt-import/css/css-masking/parsing/mask-valid.sub.txt#L53-L61

Before the rebase all 9 of these tests passed successfully. This is caused by 4c91b42dd27c14409dfb37f68e038a34d01a5eeb, reverting that commit locally fixes those tests again. @tcl3 Do you know what is going on here? Your change seems sensible to me, but the tests expect the repeated values _not_ to be shortened for the longhands. And the three major engines [all pass these tests](https://wpt.fyi/results/css/css-masking/parsing/mask-valid.sub.html?label=experimental&label=master&product=ladybird&product=chrome&product=firefox&product=safari&product=flow&product=servo).

tcl3 commented:
The shortest serialization principle I tried to follow in my PR comes from this part of the CSS specification: https://drafts.csswg.org/cssom/#serialize-a-css-value

It says:
> If component values can be omitted or replaced with a shorter representation without changing the meaning of the value, omit/replace them.

It looks like shortening repeated values in a shorthand if they occur across multiple longhands isn't valid because it would change the meaning of the value.

In addition to the tests you've highlighted, it looks like I also introduced a few regressions in: http://wpt.live/css/css-backgrounds/parsing/background-valid.html, which fails for the same reason.

One thing we could do is only shorten repeated list values if they are separated by a space (not a comma), this fixes 5 of the regressed mask subtests and all of the regressed background subtests. It doesn't seem to introduce any regressions in other parsing tests -  it does seem a bit hacky though.

Just for completeness, the change I made was (I'm not sure whether doing this is most correct fix though):
```diff
diff --git a/Libraries/LibWeb/CSS/StyleValues/StyleValueList.cpp b/Libraries/LibWeb/CSS/StyleValues/StyleValueList.cpp
index 94351b656c7..c8d3c5aba70 100644
--- a/Libraries/LibWeb/CSS/StyleValues/StyleValueList.cpp
+++ b/Libraries/LibWeb/CSS/StyleValues/StyleValueList.cpp
@@ -35,7 +35,7 @@ String StyleValueList::to_string(SerializationMode mode) const
     }

     auto first_value = m_properties.values.first();
-    if (all_of(m_properties.values, [&](auto const& property) { return property == first_value; }))
+    if (m_properties.separator == Separator::Space && all_of(m_properties.values, [&](auto const& property) { return property == first_value; }))
         return first_value->to_string(mode);

     StringBuilder builder;
```

InvalidUsernameException commented:
> It looks like shortening repeated values in a shorthand if they occur across multiple longhands isn't valid because it would change the meaning of the value.

In what scenario exactly? Both background- and mask-spec say that say that missing values are filled by repeating the existing ones. Changing `foo, foo` to `foo` just means that the single value needs to be repeated more often when using the properties for rendering. Is there an edge case I am not seeing?

> The background of a box can have multiple background image layers. **The number of layers is determined by the number of comma-separated values in the background-image property.** Note that a value of none still creates a layer.
>
> Each of the background images is sized, positioned, and tiled according to the corresponding value in the other background properties. The lists are matched up from the first value: excess values at the end are not used. **If a property doesn’t have enough comma-separated values to match the number of layers, the UA must calculate its used value by repeating the list of values until there are enough.**

https://drafts.csswg.org/css-backgrounds-3/#layering

----

> this fixes 5 of the regressed mask subtests

That is all of the regressed mask subtests ;)

AtkinsSJ commented:
> Changing `foo, foo` to `foo` just means that the single value needs to be repeated more often when using the properties for rendering. Is there an edge case I am not seeing?

It's part of what you quoted: For `background`, the number of layers is determined by how many layers are given for `background-image`, so if you deduplicated `background-image: foo, foo, foo;` and `foo` is some semi-transparent image, it would change the effect.

For the other properties that don't determine the layer count, they wouldn't have an effect, but I guess they're keep as-is for consistency.

InvalidUsernameException commented:
> so if you deduplicated `background-image: foo, foo, foo;`, it would change the effect.
Oh yeah, right. For that one specifically it matters.

Either way, since this is a regression unrelated to the changes in this PR, I think this is ready for a second review.
